app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt

```diff
@@ -449,28 +453,68 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
+	 * to create the [QueueMetadata].
 	 */
-	private fun createQueueMetadata(playerContext: PlayerContext) {
-		if (currentPlayerContext.uri != null && queueItems.isEmpty()) {
-			val listItem = ListItem(playerContext.uri, playerContext.uri, null, playerContext.title, playerContext.subtitle, false, true)
-			loadPaginatedItems(listItem, { currentPlayerContext.uri == playerContext.uri }) {
-				queueItems = removeShufflePlayButtonMetadata(it, currentPlayerContext.uri)
-
-				// builds the QueueMetadata while waiting for cover art to load
-				queueMetadata = QueueMetadata(playerContext.title, playerContext.subtitle, queueItems, mediaId = playerContext.uri)
-
-				GlobalScope.launch(defaultDispatcher) {
-					val coverArt = getQueueCoverArt()
-					queueMetadata = QueueMetadata(playerContext.title, playerContext.subtitle, queueItems, coverArt, playerContext.uri)
+	private fun createPlaylistQueueMetadata(playerContext: PlayerContext, useWebApi: Boolean) {
+		createQueueMetadataJob = GlobalScope.launch(defaultDispatcher) {
+			if (currentPlayerContext.uri != null && queueItems.isEmpty()) {
+				currentPlayerContext = playerContext
+				if (useWebApi) {
+					queueItems = webApi.getPlaylistSongs(this@SpotifyAppController, playerContext.uri) ?: emptyList()
+					if (queueItems.isNotEmpty()) {
+						createQueueMetadataFromPlayerContext(playerContext)
+					} else {
+						Log.e(TAG, "Error getting songs from playlist ${playerContext.uri}, falling back to app remote API")
+						createQueueMetadataWithAppRemote(playerContext)
+					}
+				} else {
+					createQueueMetadataWithAppRemote(playerContext)
 				}
+			}
+		}
+	}
 
-				onQueueLoaded?.invoke()
-				callback?.invoke(this)
+	/**
+	 * Creates the [QueueMetadata] for the podcast playlist.
+	 */
+	private fun createPodcastQueueMetadata(playerContext: PlayerContext) {
+		createQueueMetadataJob = GlobalScope.launch(defaultDispatcher) {
+			if (currentPlayerContext.uri != null && queueItems.isEmpty()) {
+				createQueueMetadataWithAppRemote(playerContext)
 			}
 		}
 	}
 
+	/**
+	 * Creates the [QueueMetadata] for the provided [PlayerContext] using the app remote API.
+	 */
+	private fun createQueueMetadataWithAppRemote(playerContext: PlayerContext) {
+		val listItem = ListItem(playerContext.uri, playerContext.uri, null, playerContext.title, playerContext.subtitle, false, true)
+		loadPaginatedItems(listItem, { currentPlayerContext.uri == playerContext.uri }) {
+			queueItems = removeShufflePlayButtonMetadata(it, currentPlayerContext.uri)
+			createQueueMetadataFromPlayerContext(playerContext)
+		}
+	}
+
+	/**
+	 * Creates the [QueueMetadata] for the provided [PlayerContext].
+	 */
+	private fun createQueueMetadataFromPlayerContext(playerContext: PlayerContext) {
+		// builds the QueueMetadata while waiting for cover art to load
+		queueMetadata = QueueMetadata(playerContext.title, playerContext.subtitle, queueItems, mediaId = playerContext.uri)
```

Comment:
Can `queueItems` be passed in to this function? Having this be passed as a global variable was confusing while reading through the code, and I think this might be the only use of queueItems now that you've factored the queueMetadata creation to this dedicated function, and so it might be possible to remove this instance variable entirely.

Reply:
Yeah it can be moved to a more local scope, I agree the `queueItems `and other global variables really make some of the code here confusing and actually that was something I was looking into refactoring in the long run so we can avoid these unless absolutely necessary. I removed the `queueItems ` global variable and instead have local variables handling this. We do have some queue creation that happens outside of this method as the Liked Songs and Artist playlists have more involved logic around this but I updated those helper methods to add a new queueItems parameter that will avoid our need for a global